Create naming schema section about CPE, SWID and PURL

jespunya opened the pull request:
This PR fixes #360.

- [x] This PR handles the issue and requires no additional PRs.
- [x] You have validated the need for this change.

**What did this PR accomplish?**

- Move the previous 5-Reporting/README.md to a new subsection under the 5-Reporting/0-Reporting_Structure/ folder.
- Create a new subsection of 5-Reporting at the folder document/5-Reporting/01-Naming/ with tackle naming schema with the introduction of CPE, SWID and PURL.
 
Thank you for your contribution!

kingthorin commented:
Headings should use title caps, and the section and primary index pages will require updates as well. These shouldn’t need individual directories, they should be positioned/named similar to other sections of the guide.

Was the original content just relocated or was it rewritten?

jespunya commented:
@kingthorin the original one was just moved.

kingthorin commented:
I believe the linter workflow is failing to comment due to backticks in the linting results. I'll look into fixing that separately.

kingthorin commented:
FYI if you're using VSC or Codespaces you can use the linter locally: https://github.com/OWASP/wstg/blob/master/CONTRIBUTING.md#how-to-set-up-your-contributor-environment

jespunya commented:
@kingthorin I have fixed several lint errors. Others I don't have a clear idea about how to fix them. 

On the other hand, I have fixed the files/folder structure as you mentioned & I have modified the original documentation (actually on the file 0-Reporting_Structure.md) to fix the mistakes detected by github actions (commit 27e420c).

jespunya commented:
@kingthorin Thanks for the VSC extension recommendation. All the lint errors had been fixed.

kingthorin commented on document/5-Reporting/1-Naming_Schemes.md:
```diff
@@ -0,0 +1,121 @@
+# Vulnerability naming schemes
```
```suggestion
# Vulnerability Naming Schemes
```

kingthorin commented on document/5-Reporting/1-Naming_Schemes.md:
```diff
@@ -0,0 +1,121 @@
+# Vulnerability naming schemes
+
+With a constantly growing number of IT assets to administrate, the security practitioners had required new and more powerful tools to perform automated and large scale analysis. Thanks to those software the attention can be focused on the more creative and intellectually challenging problems. Unfortunately, allowing the vulnerability assessment tools, antivirus software and intrusion detection sy ...
```
```suggestion
With a constantly growing number of IT assets to administer, security practitioners require new and more powerful tools to perform automated and large scale analysis. Thanks to software attention can be focused on the more creative and intellectually challenging problems. Unfortunately, having vulnerability assessment tools, antivirus software, and intrusion detection systems communicate its not an easy job. It has resulted in several technical complications, requiring a standardized way to identify each software flaw, vulnerability, or configuration issues identified. The lack of this interoperability capabilities can cause inconsistencies during the security assessment, confusing reporting, and extra correlation efforts among other problems that will produce an important waste of resources and time.
```

kingthorin commented on document/5-Reporting/1-Naming_Schemes.md:
```diff
@@ -0,0 +1,121 @@
+
+With a constantly growing number of IT assets to administrate, the security practitioners had required new and more powerful tools to perform automated and large scale analysis. Thanks to those software the attention can be focused on the more creative and intellectually challenging problems. Unfortunately, allowing the vulnerability assessment tools, antivirus software and intrusion detection sy ...
+
+A naming scheme is a systematic methodology used to identify each one of those vulnerabilities in order to facilitate clear identification and information shearing. This goal is achieved by the definition of a unique, structured and software-efficient name to each vulnerability. There are multiple schemes used to succeeded on this effort, the more common and used ones are:
```
```suggestion
A naming scheme is a systematic methodology used to identify each one of those vulnerabilities in order to facilitate clear identification and information sharing. This goal is achieved by the definition of a unique, structured, and software-efficient name for each vulnerability. There are multiple schemes used to facilitate this effort, the most common are:
```

kingthorin commented on document/5-Reporting/1-Naming_Schemes.md:
```diff
@@ -0,0 +1,121 @@
+
+## Software Identification Tag (SWID)
+
+Software Identification Tag (`SWID`) is an International Organization for Standardization's standard defined by the ISO/IEC 19770-2:2015. It is used primarily to identify installed software and is the preferred format of the NVD in front of `CPE`.
```
```suggestion
Software Identification Tag (`SWID`) is an International Organization for Standardization's standard defined by in ISO/IEC 19770-2:2015. It is used primarily to identify installed software and is the preferred format of the NVD in front of `CPE`.
```


> preferred format of the NVD 

"...it is NIST's perferred format for the National Vulnerability Database (NVD) over the previous standard `CPE`." ? Or something like that? (So that NVD is spelled out as it's used a few times here, and that it's clear NIST is the authority which maintains it. :man_shrugging: )

jespunya replied:
I have pushed a new text to replace this sentence to explain the relation between `swid` and `cpe`. I hope this new text will clarify your questions. If it is so, I would flag this as resolved.

kingthorin commented on document/5-Reporting/1-Naming_Schemes.md:
```diff
@@ -0,0 +1,121 @@
+
+Software Identification Tag (`SWID`) is an International Organization for Standardization's standard defined by the ISO/IEC 19770-2:2015. It is used primarily to identify installed software and is the preferred format of the NVD in front of `CPE`.
+
+Each `SWID` tag is represented as a standardized XML format. A `SWID` tag is composed for three groups of elements. The first block composed by 7 predefined elements required to be considered a valid tag. The optional block provides a set of 30 possible predefined elements that the tag creator can use to provide reliable and detailed information. Finally, the `Extended` group of elements provides ...
```
```suggestion
Each `SWID` tag is represented as a standardized XML format. A `SWID` tag is composed for three groups of elements. The first block composed by 7 predefined elements required to be considered a valid tag. Followed by an optional block which provides a set of 30 possible predefined elements that the tag creator can use to provide reliable and detailed information. Finally, the `Extended` group of elements provides the opportunity for the tag creator to define any non-predefined elements required to accurately define the described the asset. The high level of granularity provided by `SWID`, not only provides the capability to describe a given product of software, but also it's specific status on the software lifecycle.
```

I haven't looked this up so please make sure the edits maintain the accuracy of the description.

> define the described the asset

Would "software", "product", or something else be a better fit than "asset"?

jespunya replied:
I used the word "asset" to avoid the repetition of the "software" too many times, but you are right that "software" is more accurate and should be use on this case.

jespunya replied:
The corrections from your suggested change had been added on the commit `817fbcdcdcafa17d8ff491352efb7e4c77ea725e`. I flag this conversation as resolved.

kingthorin commented on document/5-Reporting/1-Naming_Schemes.md:
```diff
@@ -0,0 +1,121 @@
+
+`<SoftwareIdentity xmlns="http://standards.iso.org/iso/19770/-2/2015/schema.xsd" name="ACME Roadrunner Service Pack 1" tagId="com.acme.rms-ce-sp1-v1-0-0" patch="true" version="1.0.0"><Entity name="The ACME Corporation" regid="acme.com" role="tagCreator softwareCreator"/><Link rel="patches" href="swid:com.acme.rms-ce-v4-1-5-0"> ...</SoftwareIdentity>`
+
+- Red Hat Enterprise Linux version 8 for x86-64 architecutre:
```
```suggestion
- Red Hat Enterprise Linux version 8 for x86-64 architecture:
```

kingthorin commented on document/5-Reporting/1-Naming_Schemes.md:
```diff
@@ -0,0 +1,121 @@
+
+## Common Platform Enumeration (CPE)
+
+The Common Platform Enumperation scheme (CPE) is a structured naming scheme for information technology systems, software, and packages maintained by `NVD`. Commonly used in conjunction with the Common Vulnerabilities and Exposures identification codes (e.g. `CVE‑2017‑0147`). Despite being considered a deprecated scheme in benefit of `SWID`, `CPE` is still widely used by several security solutions ...
```
```suggestion
The Common Platform Enumeration scheme (`CPE`) is a structured naming scheme for information technology systems, software, and packages maintained by `NVD`. Commonly used in conjunction with the Common Vulnerabilities and Exposures identification codes (e.g. `CVE‑2017‑0147`). Despite being considered a deprecated scheme superseded by `SWID`, `CPE` is still widely used by several security solutions.
```

kingthorin commented on document/5-Reporting/1-Naming_Schemes.md:
```diff
@@ -0,0 +1,121 @@
+
+Defined as a Dictionary of registered values provided by `NVD`. Each `CPE` code can be defined as a well-formatted name or as a URL. Each value MUST follow this structure:
+
+- _cpe-name_ = "cpe:/" component-list
+- _component-list_ = part ":" vendor ":" product ":" version ":" update ":" edition ":" lang
+- _component-list_ /= part ":" vendor ":" product ":" version ":" update ":" edition
+- _component-list_ /= part ":" vendor ":" product ":" version ":" update
+- _component-list_ /= part ":" vendor ":" product ":" version
+- _component-list_ /= part ":" vendor ":" product
+- _component-list_ /= part ":" vendor
+- _component-list_ /= part
+- _component-list_ /= empty
+- _part_ = "h"/"o"/"a" = string
+- _vendor_ = string
+- _product_ = string
+- _version_ = string
+- _update_ = string
+- _edition_ = string
+- _lang_ LANGTAG / empty
+- _string_ = *( unreserved / pct-encoded )
+- _empty_ = ""
+- _unreserved_ = ALPHA / DIGIT / "-" / "." / "_" / " ̃"
+- _pct-encoded_ = "%" HEXDIG HEXDIG
+- _ALPHA_ = %x41-5a / %x61-7a ; A-Z or a-z
+- _DIGIT_ = %x30-39 ; 0-9
+- _HEXDIG_ = DIGIT / "a" / "b" / "c" / "d" / "e" / "f"
+- _LANGTAG_ = cf. [RFC5646]
```
Do you think it might make more sense to have this as a code block with triple backticks?

kingthorin commented on document/5-Reporting/1-Naming_Schemes.md:
```diff
@@ -0,0 +1,121 @@
+
+## Package URL (PURL)
+
+Package URL standardizes how software package metadata is represented so that packages can universally be located regardless of what vendor, project, or ecosystem the packages belongs.
```
```suggestion
Package URL standardizes how software package metadata is represented so that packages can be universally located regardless of what vendor, project, or ecosystem the packages belongs.
```

kingthorin commented on document/5-Reporting/1-Naming_Schemes.md:
```diff
@@ -0,0 +1,121 @@
+
+Package URL standardizes how software package metadata is represented so that packages can universally be located regardless of what vendor, project, or ecosystem the packages belongs.
+
+A purl is a valid `RFC3986` ASCII string defnid URL composed seven elements. Each of them is separated by a defined character in order to make it easily manipulated by software.
```
Should "PURL" generally appear in uppercase? Is there a standard we should adhere to?

kingthorin commented on document/5-Reporting/1-Naming_Schemes.md:
```diff
@@ -0,0 +1,121 @@
+
+The defintion for each components is:
+
+- _scheme_: URL scheme compilant constant value of "pkg". (**Required**).
+
+- _type_: package type or package protocol such as maven, npm, nuget, gem, pypi, etc. (**Required**).
+
+- _namespace_: type-specific value to a package prefix such as it's owner name, groupid, etc. (Optional).
+
+- _name_: name of the package. (**Required**).
+
+- _version_: package version. (Optional).
+
+- _qualifiers_: extra qualifying data for a package such as an OS, architecture, a distro, etc. (Optional).
+
+- _subpath_: extra subpath within a package, relative to the package root. (Optional).
```
I don't think this list needs to be double spaced. Other's don't seem to be.

kingthorin commented on document/5-Reporting/1-Naming_Schemes.md:
```diff
@@ -0,0 +1,121 @@
+
+`scheme:type/namespace/name@version?qualifiers#subpath`
+
+The defintion for each components is:
```
```suggestion
The definition for each component is:
```

kingthorin commented on document/5-Reporting/1-Naming_Schemes.md:
```diff
@@ -0,0 +1,121 @@
+- Curl software, packaged as a .deb Debian Jessie package for an i386 architecture: `pkg:deb/debian/curl@7.50.3-1?arch=i386&distro=jessie`
+- Docker image of Apache Casandra signed with the SHA256 hash 244fd47e07d1004f0aed9c: `pkg:docker/cassandra@sha256:244fd47e07d1004f0aed9c`
+
+### Known implementations
```
```suggestion
### Known Implementations
```

kingthorin commented on document/5-Reporting/1-Naming_Schemes.md:
```diff
@@ -0,0 +1,121 @@
+
+### Known implementations
+
+- _Golang_: <https://github.com/package-url/packageurl-go>
+- _.NET_: <https://github.com/package-url/packageurl-dotnet>
+- _JVM_: <https://github.com/package-url/packageurl-java>, <https://github.com/sonatype/package-url-java>
+- _Python_: <https://github.com/package-url/packageurl-python>
+- _Rust_: <https://github.com/package-url/packageurl-rs>
+- _JS_: <https://github.com/package-url/packageurl-js>
```
```suggestion
- _Golang_: <https://github.com/package-url/packageurl-go>
- _.NET_: <https://github.com/package-url/packageurl-dotnet>
- [packageurl-java](https://github.com/package-url/packageurl-java], [package-url-java](https://github.com/sonatype/package-url-java)
- [Python](https://github.com/package-url/packageurl-python)
- [Rust](https://github.com/package-url/packageurl-rs)
- [JavaScript](https://github.com/package-url/packageurl-js)
```

kingthorin replied:
Maybe move this to the references section?

kingthorin commented on document/5-Reporting/1-Naming_Schemes.md:
```diff
@@ -0,0 +1,121 @@
+- _Rust_: <https://github.com/package-url/packageurl-rs>
+- _JS_: <https://github.com/package-url/packageurl-js>
+
+## Recommendation usages
```
```suggestion
## Recommendation Uses
```

kingthorin commented on document/5-Reporting/1-Naming_Schemes.md:
```diff
@@ -0,0 +1,121 @@
+| Operating System | CPE or SWID |
+| Operating System Package | PURL or SWID |
+
+_Note:_ Due to the deprecated status of `CPE`, it's recomended that the new projects implement `SWID` when they need to decide between both methods. Even though, `CPE` is known to be a widely used naming schema on currently active projects.
```
```suggestion
> Note: Due to the deprecated status of `CPE`, industry recommended seems to be that the new projects implement `SWID` when they need to decide between the two methods. Even though `CPE` is known to be a widely used naming schema within current active projects and solutions.
```

> between the two methods

Would "standards" or "schemes" be better than "methods"?

kingthorin commented on document/5-Reporting/1-Naming_Schemes.md:
```diff
@@ -0,0 +1,121 @@
+
+_Note:_ Due to the deprecated status of `CPE`, it's recomended that the new projects implement `SWID` when they need to decide between both methods. Even though, `CPE` is known to be a widely used naming schema on currently active projects.
+
+## Additional information
```
```suggestion
## References
```

To be consistent with the [template](https://github.com/OWASP/wstg/blob/master/template/999-Foo_Testing/2-Template_Explanation.md#references)

kingthorin commented on document/5-Reporting/1-Naming_Schemes.md:
```diff
@@ -0,0 +1,121 @@
+
+## Additional information
+
+- [NISTIR 8060 - Guidelines for the Creation of Interoperable Software Identification (SWID) Tags](https://nvlpubs.nist.gov/nistpubs/ir/2016/NIST.IR.8060.pdf)
```
```suggestion
- [NISTIR 8060 - Guidelines for the Creation of Interoperable Software Identification (SWID) Tags (pdf)](https://nvlpubs.nist.gov/nistpubs/ir/2016/NIST.IR.8060.pdf)
```

kingthorin commented on document/5-Reporting/README.md:
```suggestion
5.2 [Naming Schemes](1-Naming_Schemes.md)
```

The link text needs to be consistent with the file name. (The file name should be `2-*` I believe)

jespunya replied:
I think not, other sections like for example: `3-The_OWASP_Testing_Framework` starts from `0-*` too. It's subsections are: `3-The_OWASP_Testing_Framework/0-The_Web_Security_Testing_Framework.md` & `3-The_OWASP_Testing_Framework/1-Penetration_Testing_Methodologies.md`.

kingthorin replied:
The link text absolutely need to align with the file name, otherwise links in the generated PDF won’t work

kingthorin commented on document/5-Reporting/README.md:
```diff
@@ -1,126 +1,5 @@
 # Reporting
 
-Performing the technical side of the assessment is only half of the overall assessment process. The final product is the production of a well written and informative report. A report should be easy to understand and should highlight all the risks found during the assessment phase. The report should appeal to both executive management and technical staff.
+5.1 [Reporting Structure](0-Reporting_Structure.md)
```
```suggestion
5.1 [Reporting Structure](0-Reporting_Structure.md)
```

The filename should be `1-*` I believe.

jespunya replied:
I think not, other sections like for example: `3-The_OWASP_Testing_Framework` starts from `0-*` too. It's subsections are: `3-The_OWASP_Testing_Framework/0-The_Web_Security_Testing_Framework.md` & `3-The_OWASP_Testing_Framework/1-Penetration_Testing_Methodologies.md`.

kingthorin replied:
As with section 3, the readme should become a section ToC. The current readme content should become the 0th document in the section.

kingthorin commented on document/5-Reporting/0-Reporting_Structure.md:
```diff
@@ -0,0 +1,171 @@
+# Reporting
```
I haven't really reviewed this content as I expect it to be unchanged. I'll diff it at some point before merge.

jespunya replied:
It was not correct since was not updated with the latest changes done on master, I have just updated it on commit `227959927882d4e349d1544c509727d0861b9675` .

kingthorin reviewed (changes requested):
Thanks for tackling this.

https://github.com/OWASP/wstg/blob/master/document/README.md should also be updated.

kingthorin commented:
@jespunya are you able to address the latest feedback or do you need someone else to finish this?

jespunya commented:
> **Following links are broken:**
> FILE:document/4-Web_Application_Security_Testing/09-Testing_for_Weak_Cryptography/03-Testing_for_Sensitive_Information_Sent_via_Unencrypted_Channels.md
> [heavy_multiplication_x] https://owasp.org/www-community/attacks/Man-in-the-middle_attack → Status: 404

I have created the issue: https://github.com/OWASP/wstg/issues/736 where I will fix it as an independent PR.

jespunya commented:
> > **Following links are broken:**
> > FILE:document/4-Web_Application_Security_Testing/09-Testing_for_Weak_Cryptography/03-Testing_for_Sensitive_Information_Sent_via_Unencrypted_Channels.md
> > [heavy_multiplication_x] https://owasp.org/www-community/attacks/Man-in-the-middle_attack → Status: 404
> 
> I have created the issue: #736 where I will fix it as an independent PR.

kingthorin commented:
I've rebased this so that it doesn't include a bunch of unrelated commits and file changes. I'll do a full review in a few hours (and provide instructions on how you should update your local). [Please don't make or push any changes in the meantime.]

kingthorin commented:
I decided to just fix the remaining items instead of marking it up again.

Please do the following to get your local copy in-sync with these changes (assuming you're in the wstg repo directory in your computer, and that your fork is set as `origin`):
```console
git fetch origin
git reset --hard origin/master
```

victoriadrake commented:
This one should wait for v5, yeah?

kingthorin commented:
Hmmm it doesn’t actually impact on testing scenarios so o think we can add it now.

jespunya commented:
The related issue: https://github.com/OWASP/wstg/issues/360 was flaged on November to be part of the v5 as @ThunderSon proposed previously on the issue itself. Even though, for me both options would be ok.

victoriadrake commented:
Just waiting for v5.

kingthorin commented:
Does this really need to wait? (@ThunderSon)

kingthorin commented:
Poke @ThunderSon

ThunderSon commented:
Apologies on this. Are we certain on the reporting page? Is this happening because of the `master` branch? 🤔 
I believe this isn't breaking any previous versions, so it should be good.

kingthorin commented on document/5-Reporting/0-Reporting_Structure.md:
```diff
@@ -0,0 +1,126 @@
+# Reporting
```
This file should be named 1- not 0-

kingthorin commented on document/5-Reporting/1-Naming_Schemes.md:
```diff
@@ -0,0 +1,121 @@
+# Vulnerability Naming Schemes
```
This file should be 2- not 1-

kingthorin reviewed (commented):
Once the file names are adjusted to match the proper section references the indexes will need to be tweaked

ThunderSon commented:
Why should they be changed? We are starting our chapters from the 0 numbering. How are you looking at this?

kingthorin commented:
There are only 0th items in chapters that have generic lead-ins. Ex: No 0th item: https://github.com/OWASP/wstg/tree/master/document/4-Web_Application_Security_Testing/01-Information_Gathering

Also 5.1 would never link a 0th page: `5.1 [Reporting Structure](0-Reporting_Structure.md)`

Actually I believe the only place we actually have a 0th page is in chapter 3.

ThunderSon commented:
I agree. Should we update chapter 3? I believe that'd break compatibility between v42 and v43 when it's released.
Can you please do the changes needed on this PR? Let's get this in :)

kingthorin commented:
I guess I'll have to pull this on my dev box and rebase it. Will do that shortly.

kingthorin commented:
@ThunderSon I believe this is ready now.

kingthorin commented:
@ThunderSon this one good to go?

ThunderSon reviewed (approved):
@kingthorin all yours!

kingthorin commented:
Thank you both!
